autocorns/biologist.py

```diff
@@ -98,149 +127,137 @@ def unicorn_metadata(
             result = {
                 "token_id": token_id,
                 "block_number": block_number,
-                "lifecycle_stage": metadata[3],
-                "class_number": metadata[-2],
+                "lifecycle_stage": token_data[3],
+                "class_number": token_data[-2],
             }
             results.append(result)
         except Exception as e:
             error = {
                 "token_id": token_id,
                 "block_number": block_number,
-                "error": f"Failed to retrieve DNA: {str(e)}",
+                "error": f"Failed retrive unicorns metadata: {str(e)}",
             }
             errors.append(error)
-
     return results, errors
 
 
 def unicorn_mythic_body_parts(
     contract_address: ChecksumAddress,
     dnas: List[Dict[str, Any]],
     block_number: Optional[int] = None,
-    num_workers: int = 1,
-    timeout: float = 30.0,
-    checkpoint_file: Optional[str] = None,
 ) -> Tuple[List[Dict[str, Any]], List[Dict[str, Any]]]:
     if block_number is None:
         block_number = len(chain) - 1
 
     results: List[Dict[str, Any]] = []
     errors: List[Dict[str, Any]] = []
 
-    dnas_index = {item["token_id"]: item for item in dnas}
-
-    checkpointed_token_ids: Set[int] = set()
-    if checkpoint_file is not None:
-        with open(checkpoint_file, "r") as ifp:
-            for line in ifp:
-                stripped_line = line.strip()
-                if stripped_line:
-                    result = json.loads(stripped_line)
-                    if result["token_id"] not in dnas_index:
-                        continue
-                    dna_result = dnas_index[result["token_id"]]
-                    if (
-                        result.get("num_mythic_body_parts") is not None
-                        and result.get("dna") != "2"
-                        and dna_result["block_number"] <= result["block_number"]
-                    ):
-                        results.append(result)
-                        checkpointed_token_ids.add(result["token_id"])
-
-    submission_progress_bar = tqdm(
-        total=len(dnas) - len(checkpointed_token_ids),
-        desc="Submitting requests for number of unicorn mythic body parts",
-    )
+    tokens_metadata = []
 
     mythic_progress_bar = tqdm(
-        total=len(dnas) - len(checkpointed_token_ids),
+        total=len(dnas),
         desc="Retrieving number of unicorn mythic body parts",
     )
 
     contract = StatsFacet.StatsFacet(contract_address)
 
-    jobs: Dict[Future, int] = {}
-    with ThreadPoolExecutor(max_workers=num_workers) as executor:
-        for item in dnas:
-            if item["token_id"] in checkpointed_token_ids:
+    block_number = dnas[0]["block_number"]
+
+    multicaller = Multicall2.Multicall2(Multicall2_address)
+
+    multicall_method = multicaller.contract.tryAggregate
+
+    dnas_is_present = [
+        dna for dna in dnas if dna["dna"] is not None and dna["dna"] != "None"
```

Comment:
dna.get("dna") better than `dna["dna"]`?

Reply:
Is token ID being represented here?

Reply:
Skip for now maybe return later.